implementations/rust/ockam/ockam_api/src/verifier.rs

```diff
@@ -95,27 +94,22 @@ where
         let data = CredentialData::try_from(cre)?;
 
         let ident = if let Some(ident) = req.authority(data.unverfied_issuer()) {
-            IdentityChangeHistory::import(ident)?
+            PublicIdentity::import(ident, &self.vault).await?
         } else {
             let err = Error::new("/verify").with_message("unauthorised issuer");
             return Ok(Either::Left(Response::unauthorized(id).body(err)));
         };
 
-        let data = cre.verify_signature(&ident, &self.vault).await?;
-
-        if req.subject() != data.subject() {
-            let err = Error::new("/verify").with_message("invalid subject");
-            return Ok(Either::Left(Response::forbidden(id).body(err)));
-        }
-
-        let now = Timestamp::now().ok_or_else(|| {
-            ockam_core::Error::new(Origin::Core, Kind::Internal, "invalid system time")
-        })?;
-
-        if now >= data.expires_at() {
-            let err = Error::new("/verify").with_message("expired");
-            return Ok(Either::Left(Response::forbidden(id).body(err)));
-        }
+        let data = match ident
+            .verify_credential(cre, req.subject(), &self.vault)
+            .await
+        {
+            Ok(data) => data,
+            Err(_) => {
```

Comment:
Would be helpful to preserve the error information (at least in logs).

The whole method can be simplified somewhat:


```rust
async fn verify<'a>(
    &self,
    id: Id,
    req: &'a VerifyRequest<'a>,
    cre: &'a Credential<'a>,
) -> Result<Either<ResponseBuilder<Error<'_>>, CredentialData<'a, Verified>>> {
    let data = CredentialData::try_from(cre)?;

    if let Some(ident) = req.authority(data.unverfied_issuer()) {
        let ident = PublicIdentity::import(ident, &self.vault).await?;
        let data = ident.verify_credential(cre, req.subject(), &self.vault).await?;
        Ok(Either::Right(data))
    } else {
        let err = Error::new("/verify").with_message("unauthorised issuer");
        Ok(Either::Left(Response::unauthorized(id).body(err)))
    }
}
```

Reply:
@twittner the difference is that I have 2 different error messages returned to the client depending if authority wasn't found in the list or verification failed. No big deal if it's one error, but I wouldn't call that "simplification". Therefore, I would leave as is

Reply:
At least log the error message.

Reply:
@twittner I think we want to develop an approach for better error handling in our Services overall. My suggestion is that there should be a clear boundary between following:
1. Serving a service (deserializing requests, responding with responses) 
1. Performing an action caused by the request
 
In this case `2` should operate using ockam rust errors, while `1` should return meaningful error in case of:
1. Invalid requests
1. Transforming underlying error `1` returned wrapped in a proper Response

This should happen across whole codebase, maybe some changes to rust error should also happen to make this possible. Until then, errors we return in services are a mess and will be a mess. But, if you think propagating error in this specific function is an improvement worth discussion - I'll push the change

Reply:
I agree with the distinction and I am not against the generic error response message, but once an error response is received it may be necessary to analyse logs to understand what went wrong and seeing a potentially more detailed error in the logs can save a lot of time. Regardless of where the code is right now I think errors should always be logged, unless there is a good reason not to (e.g. because of sensitive information).